File: backends/java/zappy-adapters/src/main/java/com/zappymart/adapters/security/Argon2PasswordHasher.java
package com.zappymart.adapters.security;

import com.zappymart.application.ports.PasswordHasher;
import org.springframework.security.crypto.argon2.Argon2PasswordEncoder;

public final class Argon2PasswordHasher implements PasswordHasher {

    public static final int SALT_LENGTH_IN_BYTES = 16;

    public static final int HASH_LENGTH_IN_BYTES = 32;

    public static final int PARALLELISM = 1;

    public static final int MEMORY_IN_KIBIBYTES = 19456;

    public static final int ITERATIONS = 2;

    private final Argon2PasswordEncoder encoder = new Argon2PasswordEncoder(
            SALT_LENGTH_IN_BYTES, HASH_LENGTH_IN_BYTES, PARALLELISM, MEMORY_IN_KIBIBYTES, ITERATIONS);

    @Override
    public String hash(String password) {
        return encoder.encode(password);
    }

    @Override
    public boolean matches(String password, String storedHash) {
        return encoder.matches(password, storedHash);
    }
}

File: backends/kotlin/zappy-adapters/src/main/kotlin/nl/zappymart/adapters/security/Argon2PasswordHasher.kt
package nl.zappymart.adapters.security

import nl.zappymart.application.ports.PasswordHasher
import nl.zappymart.domain.accounts.PasswordHash
import org.springframework.security.crypto.argon2.Argon2PasswordEncoder
import org.springframework.stereotype.Component

@Component
class Argon2PasswordHasher : PasswordHasher {

    private val encoder = Argon2PasswordEncoder(
        SALT_LENGTH_IN_BYTES,
        HASH_LENGTH_IN_BYTES,
        PARALLELISM,
        MEMORY_IN_KIBIBYTES,
        ITERATIONS,
    )

    override fun hash(password: String) = PasswordHash(requireNotNull(encoder.encode(password)))

    override fun matches(password: String, hash: PasswordHash) = encoder.matches(password, hash.value)

    private companion object {
        const val SALT_LENGTH_IN_BYTES = 16
        const val HASH_LENGTH_IN_BYTES = 32
        const val PARALLELISM = 1
        const val MEMORY_IN_KIBIBYTES = 19456
        const val ITERATIONS = 2
    }
}
